XX_BubbleSort/main.cpp

```diff
@@ -0,0 +1,449 @@
+		// The Logical Device itself implements the interface and behaves as the most simple allocator, it will create a new `nbl::video::IDeviceMemoryAllocation` every single time.
+		// We will cover allocators and suballocation in a later example.
+		nbl::video::IDeviceMemoryAllocator::SAllocation allocation = {};
+
+		// For our Compute Shader
+		constexpr uint32_t WorkgroupSize = 256;
+		constexpr uint32_t WorkgroupCount = 2048;
+
+		// This scope is kinda silly but it demonstrated that in Nabla we refcount all Vulkan resources and keep the ones used in sumbits (semaphores and commandbuffers) alive till the submit is no longer pending
+		{
+			// You should already know Vulkan and come here to save on the boilerplate, if you don't know what instances and instance extensions are, then find out.
+			smart_refctd_ptr<nbl::video::CVulkanConnection> api;
+			{
+				// You generally want to default initialize any parameter structs
+				nbl::video::IAPIConnection::SFeatures apiFeaturesToEnable = {};
+				// generally you want to make your life easier during development
+				apiFeaturesToEnable.validations = true;
+				apiFeaturesToEnable.synchronizationValidation = true;
+				// want to make sure we have this so we can name resources for vieweing in RenderDoc captures
+				apiFeaturesToEnable.debugUtils = true;
+				// create our Vulkan instance
+				if (!(api = CVulkanConnection::create(smart_refctd_ptr(m_system), 0, _NBL_APP_NAME_, smart_refctd_ptr(base_t::m_logger), apiFeaturesToEnable)))
+					return logFail("Failed to crate an IAPIConnection!");
+			}
+
+			// We won't go deep into performing physical device selection in this example, we'll take any device with a compute queue.
+			uint8_t queueFamily;
+			// Nabla has its own set of required baseline Vulkan features anyway, it won't report any device that doesn't meet them.
+			nbl::video::IPhysicalDevice* physDev = nullptr;
+			ILogicalDevice::SCreationParams params = {};
+			for (auto physDevIt = api->getPhysicalDevices().begin(); physDevIt != api->getPhysicalDevices().end(); physDevIt++)
+			{
+				const auto familyProps = (*physDevIt)->getQueueFamilyProperties();
+				// this is the only "complicated" part, we want to create a queue that supports compute pipelines
+				for (auto i = 0; i < familyProps.size(); i++)
+					if (familyProps[i].queueFlags.hasFlags(IQueue::FAMILY_FLAGS::COMPUTE_BIT))
+					{
+						physDev = *physDevIt;
+						queueFamily = i;
+						params.queueParams[queueFamily].count = 1;
+						break;
+					}
+			}
+			if (!physDev)
+				return logFail("Failed to find any Physical Devices with Compute capable Queue Families!");
+
+			// logical devices need to be created form physical devices which will actually let us create vulkan objects and use the physical device
+			device = physDev->createLogicalDevice(std::move(params));
+			if (!device)
+				return logFail("Failed to create a Logical Device!");
```

Comment:
Hello Compute was special here, because its the 2nd example, so it did more stuff "from the ground up"

You chouls use a more complex base class, like `MonoDeviceApplication` or `BasicMultiQueue`... then you can skip all this queue choice, device creation code... see example 05